os2phonebook_service/os2phonebook/controller.py

```diff
@@ -0,0 +1,316 @@
+    and carry a json body with the following key value pairs,
+
+        search_type: <string>
+        search_value: <string>
+
+    Example:
+        {
+            "search_type": "employee_by_email",
+            "search_value": "someuser@example.org"
+        }
+
+    Returns:
+        :obj:`Response`: Response with json body.
+
+    """
+
+    search_schema = {
+        "method": "POST",
+        "format": "json",
+        "schema": {
+            "search_type": {
+                "type": "string",
+                "required": True
+            },
+            "search_value": {
+                "type": "string",
+                "required": True
+            }
+        }
+    }
+
+    return jsonify(search_schema)
+
+
+@api.route("/api/search", methods=["POST"])
+def call_search_method():
+    """Perform a high level search.
+
+    Available search types:
+        * employee_by_name
+        * employee_by_phone
+        * employee_by_email
+        * employee_by_engagement
+        * org_unit_by_name
+
+    Returns:
+        :obj:`Response`: Response with json body.
+
+    """
+
+    if not request.data:
+        raise InvalidRequestBody("Request body (json) is missing")
+
+    data = request.get_json()
+
+    if "search_value" not in data:
+        raise InvalidRequestBody(
+            "Search value is missing from the request body"
+        )
+
+    if "search_type" not in data:
+        raise InvalidRequestBody(
+            "Search type is missing from the request body"
+        )
+
+    search_type = data["search_type"]
+    search_value = data["search_value"]
+
+    db = DataStore(current_app.connection)
+
+    print("SEARCH ONE")
+    results = db.search(
+        search_type=search_type,
+        search_value=search_value,
+        fuzzy_search=False
+    )
+
+    if not results:
+        log.debug(
+            f"NO_SEARCH_RESULTS search_type={search_type} search_value={search_value} # 1"
+        )
+        print("SEARCH TWO")
+        results = db.search(
+            search_type=search_type,
+            search_value=search_value,
+            fuzzy_search=True
+        )
+
+    if not results:
+        log.debug(
+            f"NO_SEARCH_RESULTS search_type={search_type} search_value={search_value} # 2"
+        )
+
+    return jsonify(results)
+
+
+#####################################################################
+#   ERROR HANDLING SECTION                                          #
+#####################################################################
+
+@api.app_errorhandler(NotFoundError)
+@api.app_errorhandler(InvalidSearchType)
+@api.app_errorhandler(InvalidRequestBody)
+def invalid_validation_handler(error) -> Response:
+    """Error handler for all common types
+
+    The supported error types are as follows:
+        * NotFoundError
+        * InvalidSearchType
+        * InvalidRequestBody
+
+    All error types above carry a `status_code`.
+    Although NotFoundError should not return 404
```

Comment:
>  Although NotFoundError should not return 404

Why not?

Reply:
You are corret, as NotFoundError will not be thrown when performing a search but ONLY when targeting a document by identifier, this should indeed return 404 as the targeted resource is not available.